runtime/vm/VMSnapshotImpl.cpp

```diff
@@ -1269,65 +1531,344 @@ resetVMThreadState(J9JavaVM *vm, J9VMThread *restoreThread)
 	restoreThread->functions = (JNINativeInterface_*) vm->jniFunctionTable;
 }
 
-extern "C" BOOLEAN
-initRestoreThreads(J9JavaVM *vm, omrthread_t mainOSThread)
+/**
+ * Restore acquired object-monitor if it matches the fixup information in monitor.
+ * 
+ * @param monitor fixup information for acquired monitor. This will be incremented to the next data structure if object-monitor is successfully restored.
+ * @param monitorCount number of monitors left to fix up. Should be decremented if monitor is successfully restored.
+ * @param object object-monitor to restore with monitor information
+ * 
+ * @return true if successful, else false
+ */
+bool
+VMSnapshotImpl::restoreAcquiredObjectMonitor(J9AcquiredMonitor** monitor, UDATA* monitorCount, j9object_t object)
 {
 	bool rc = true;
-	J9MemoryManagerFunctions* gcFuncs = vm->memoryManagerFunctions;
-	J9VMThread *restoreThread = vm->mainThread;
-	omrthread_t restoreOSThread = mainOSThread;
-	char *threadName = NULL;
+	if ((*monitorCount > 0) && ((*monitor)->isObjectMonitor) && ((*monitor)->fixupReference == (UDATA)object)) {
+		J9VMThread* monitorOwnerThread = (J9VMThread*)(*monitor)->ownerVmThreadAddress;
 
-	restoreThread->privateFlags2 |= J9_PRIVATE_FLAGS2_RESTORE_MAINTHREAD;
+		omrthread_t omrThread = monitorOwnerThread->osThread;
 
-	restoreThread->linkNext = vm->mainThread;
-	restoreThread->linkPrevious = vm->mainThread;
+		/* re-force inflation so the omrthread_monitor_t can be fixed up */
+		j9objectmonitor_t *lockEA = J9OBJECT_MONITOR_EA(monitorOwnerThread, object);
+		j9objectmonitor_t lock = J9_LOAD_LOCKWORD(monitorOwnerThread, lockEA);
 
-	resetVMThreadState(vm, restoreThread);
+		J9ObjectMonitor *objectMonitor = objectMonitorInflate(monitorOwnerThread, object, lock);
+		if (NULL == objectMonitor) {
+			rc = false;
+			goto done;
+		}
 
-	OMR_VMThread *omrVMThread = (OMR_VMThread*)(((UDATA)restoreThread) + J9_VMTHREAD_SEGREGATED_ALLOCATION_CACHE_OFFSET + vm->segregatedAllocationCacheSize);
-	memset(omrVMThread, 0, sizeof(OMR_VMThread));
+		omrthread_monitor_t omrMonitor = objectMonitor->monitor;
 
-	omrthread_monitor_init_with_name(&restoreThread->publicFlagsMutex, J9THREAD_MONITOR_JLM_TIME_STAMP_INVALIDATOR, "Thread public flags mutex");
-	if (restoreThread->publicFlagsMutex == NULL) {
-		rc = false;
-		goto done;
+		/* monitor count may be wrong since objectMonitorInflate sets it to J9_FLATLOCK_COUNT(lock).
+		 * notes for this macro indicate that the value may be incorrect for an inflated monitor.
+		 * this monitor has been previously inflated during snapshot run */
+		((J9ThreadAbstractMonitor*)omrMonitor)->count = 1;
```

Comment:
I'm unconvinced about this being correct (see comments above).  What happens if this isn't done?

Reply:
This is for the case where an object monitor was inflated at the point of snapshot so `J9_FLATLOCK_COUNT` is not accurate. The object monitor is expected to be inflated on the restore run as well. `objectMonitorInflate` sets up a system monitor for the inflated object monitor but assumes that  `J9_FLATLOCK_COUNT` is accurate.

I ran into a case during testing where `J9_FLATLOCK_COUNT` was 2 but the persisted `->count` which is saved as J9AcquiredMonitor:ownerCount was 1 which caused a hang in the program.

Reply:
`J9_FLATLOCK_COUNT` is only valid if the lock is flatlocked.  Once inflated, the `->count` must be correct.

Reply:
I've updated the comments to explain this better.